WalletWasabi/Blockchain/TransactionBuilding/BnB/CheapestSelectionStrategy.cs

```diff
@@ -0,0 +1,74 @@
+using System.Linq;
+
+namespace WalletWasabi.Blockchain.TransactionBuilding.BnB;
+
+/// <summary>
+/// Strategy that searches search-space and caches every found selection that minimizes
+/// waste of user's fund by looking for a selection that minimizes inputs' spending costs
+/// and extra cost of paying more than specified target.
+/// </summary>
+public class CheapestSelectionStrategy : ISearchStrategy
+{
+	private long _currentInputCosts = 0;
+	private long _bestTargetSoFar = long.MaxValue;
+	private long[]? _bestSelectionSoFar;
+
+	/// <param name="target">Value in satoshis.</param>
+	/// <param name="inputSpendingCosts">Costs of spending coins in satoshis.</param>
+	public CheapestSelectionStrategy(long target, long[] inputSpendingCosts)
+	{
+		Target = target;
+		InputCosts = inputSpendingCosts;
+	}
+
+	/// <inheritdoc/>
+	public long Target { get; }
+	public long[] InputCosts { get; }
+
+	/// <summary>Gives lowest found value selection whose sum is larger than or equal to <see cref="Target"/>.</summary>
+	public long[]? GetBestSelectionFound() => _bestSelectionSoFar?.Where(x => x > 0).ToArray();
+
+	/// <inheritdoc/>
+	public long UpdateSum(NextAction action, long[] selection, int depth, long oldSum)
```

Comment:
So `UpdateSum` is about updating helper variables (like keeping selection `sum` up to date). `UpdateSum` might not be great name but `UpdateEffValue` is imho worse because there is no effective value in BnB paper imho whereas BnB certainly has the concept of a sum of selected values. So *sum* seems like a good name to me.

Other names that came to my mind:

* `OnNodeExpand` / `OnNodeEnter` to denote that we call the method when we expand a node in the search space.
* `ProcessAction` ... not that bad I guess

Reply:
I have used `ProcessAction`.